public/app/features/alerting/unified/components/rules/state-history/LogTimelineViewer.tsx

```diff
@@ -0,0 +1,94 @@
+import React, { useEffect, useRef } from 'react';
+import AutoSizer from 'react-virtualized-auto-sizer';
+import { BehaviorSubject } from 'rxjs';
+
+import { DataFrame, TimeRange } from '@grafana/data';
+import { VisibilityMode } from '@grafana/schema';
+import { LegendDisplayMode, useTheme2 } from '@grafana/ui';
+import { TimelineChart } from 'app/core/components/TimelineChart/TimelineChart';
+import { TimelineMode } from 'app/core/components/TimelineChart/utils';
+
+interface LogTimelineViewerProps {
+  frames: DataFrame[];
+  timeRange: TimeRange;
+  onPointerMove?: (seriesIdx: number, pointerIdx: number) => void;
+}
+
+export const LogTimelineViewer = React.memo(({ frames, timeRange, onPointerMove }: LogTimelineViewerProps) => {
+  const theme = useTheme2();
+
+  const pointerSubject = useRef(
+    new BehaviorSubject<{ seriesIdx: number; pointIdx: number }>({ seriesIdx: 0, pointIdx: 0 })
+  );
+
+  useEffect(() => {
+    const subscription = pointerSubject.current.subscribe(({ seriesIdx, pointIdx }) => {
+      onPointerMove && onPointerMove(seriesIdx, pointIdx);
+    });
+
+    return () => {
+      subscription.unsubscribe();
+    };
+  }, [onPointerMove]);
+
+  return (
+    <AutoSizer disableHeight>
+      {({ width }) => (
+        <TimelineChart
+          frames={frames}
+          timeRange={timeRange}
+          timeZone={'browser'}
+          mode={TimelineMode.Changes}
+          height={18 * frames.length + 50}
+          width={width}
+          showValue={VisibilityMode.Never}
+          theme={theme}
+          rowHeight={0.8}
+          legend={{
+            calcs: [],
+            displayMode: LegendDisplayMode.List,
+            placement: 'bottom',
+            showLegend: true,
+          }}
+          legendItems={[
+            { label: 'Normal', color: theme.colors.success.main, yAxis: 1 },
+            { label: 'Pending', color: theme.colors.warning.main, yAxis: 1 },
+            { label: 'Alerting', color: theme.colors.error.main, yAxis: 1 },
+            { label: 'NoData', color: theme.colors.info.main, yAxis: 1 },
+          ]}
+        >
+          {(builder) => {
+            builder.setSync();
```

Comment:
My think my largest concern in this PR is this part (from a legibility perspective). 

Can we abstract this in to a separate function so we can

1. keep the perceived component complexity lower here
2. make more room for code comments (there's an opportunity here to document this complex interaction for future use-cases).

Reply:
I extracted the code responsible for tracking cursor movement to a separate hook. I think it looks much better now!